### PR TITLE
Write attributes on file top level

### PR DESCRIPTION
**Pull request by dominikwerder**
Write standard top-level HDF attributes, like creation time/date, filename, HDF version, into the output HDF file.

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
I've removed reference to the `NeXus_version` attribute from the PR description because it is not implemented in the changes. According to the NeXus documentation it should "Only used when the NAPI has written the file" anyway.

File: src/tests/HDFFile.cxx [hunk 1]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
     H5Fclose(fid);
   }
 
+  static void write_attributes_at_top_level_of_the_file() {
+    using namespace FileWriter;
+    using std::array;
+    using std::vector;
+    using std::string;
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: These `using std`s can be removed.

File: src/tests/HDFFile.cxx [hunk 2]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
...
+    MainOpt &main_opt = *g_main_opt.load();
+    {
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: Same comment as in #98, this block can be a reusable function

File: src/tests/HDFFile.cxx [hunk 3]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
...
+    })"");
+    auto cmd = json_to_string(json_command);
+    auto fname = get_string(&json_command, "file_attributes.file_name");
+    ASSERT_GT(fname.v.size(), 8);
+
+    FileWriter::CommandHandler ch(main_opt, nullptr);
+    Msg msg;
+    msg.data = (char *)cmd.data();
+    msg.size = cmd.size();
+    ch.handle(msg);
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: Could use `ch.handle({(char *)cmd.data(), cmd.size()})` instead.

File: src/tests/HDFFile.cxx [hunk 4]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
...
+    ASSERT_EQ(ch.file_writer_tasks.size(), (size_t)1);
+    {
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: Can use the same `job_id` as in the other tests, then this is also a repeated code block with can be refactored to a separate function.

File: src/tests/HDFFile.cxx [hunk 5]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
...
+        "cmd": "file_writer_tasks_clear_all",
+        "job_id": "832yhtwgfskdf"
+      })"");
+      ch.handle({(char *)cmd.data(), cmd.size()});
+    }
+
+    // Verification
+    auto fid = H5Fopen(string(fname).c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);
+    ASSERT_GE(fid, 0);
+    {
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: I would also suggest putting these in separate functions to improve readability, they could also be useful in other tests.

File: src/tests/HDFFile.cxx [hunk 6]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
...
+      auto dtype = H5Aget_type(attr);
+      ASSERT_EQ(H5Tget_class(dtype), H5T_STRING);
+      std::vector<char> buf;
+      buf.resize(128);
+      H5Aread(attr, dtype, buf.data());
+      buf[buf.size() - 1] = 0;
+      ASSERT_EQ(string("Hello Attribute"), buf.data());
+      H5Aclose(attr);
+    }
+    {
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: If this was a function it could also be used to check the `file_name` attribute.

File: src/tests/HDFFile.cxx [hunk 7]
```diff
@@ -286,6 +286,91 @@ class T_CommandHandler : public testing::Test {
...
+      ASSERT_EQ(H5Tget_class(dtype), H5T_STRING);
+      auto dtype_mem = H5Tget_native_type(dtype, H5T_DIR_ASCEND);
+      std::vector<char> buf;
+      buf.resize(128);
+      H5Aread(attr, dtype_mem, buf.data());
+      buf[buf.size() - 1] = 0;
+      ASSERT_EQ(h5_version_string_linked(), buf.data());
+      H5Aclose(attr);
+    }
+    H5Fclose(fid);
```
Comment (matthew-d-jones) [inline src/tests/HDFFile.cxx]: Should `file_time` always be present? If so, you could test that there is an attribute with that name present.

File: src/tests/HDFFile.cxx (hunk 6 above)
Comment (dominikwerder) [reply src/tests/HDFFile.cxx]: True, there is quite some opportunity for factoring here, but I keep that for later.

**Pull request merged**